src/Storages/StorageS3.cpp

```diff
@@ -182,24 +182,21 @@ String StorageS3Source::DisclosedGlobIterator::next()
 class StorageS3Source::KeysIterator::Impl
 {
 public:
-    explicit Impl(const std::vector<String> & keys_) : keys(keys_), keys_iter(keys.begin())
+    explicit Impl(const std::vector<String> & keys_) : keys(keys_)
     {
     }
 
     String next()
     {
-        std::lock_guard lock(mutex);
-        if (keys_iter == keys.end())
+        size_t current_index = index.fetch_add(1);
+        if (current_index >= keys.size())
             return "";
-        auto key = *keys_iter;
-        ++keys_iter;
-        return key;
+        return keys[current_index];
```

Comment:
These changes are not related to bug, just decided to use atomic instead of mutex here